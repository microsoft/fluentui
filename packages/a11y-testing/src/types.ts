// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type PropValue = string | number | boolean | any;
export type Props = { [name: string]: PropValue };
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AccessibilityBehavior = (props: Props) => any;

export interface TestFacade {
  slotExists(slotName: string): boolean;
  attributeExists(slotName: string, attributeName: string): boolean;
  attributeHasValue(slotName: string, attributeName: string, value: PropValue): boolean;
  getAttributeValue(slotName: string, attribute: string): PropValue;
  forProps(props: Props): TestFacade;
  afterClick(slotName: string): void;
  pressSpaceKey(slotName: string): void;
  pressEnterKey(slotName: string): void;
  verifyOnclickExecution(slotName: string): boolean;
}

export interface Rule {
  stringify: () => string;
}

export interface Slot {
  name: string;
  props: Props[];
  expectAttribute: boolean;
  expectedAttribute: string;
  expectedValue: PropValue;
  description: string;
  hidden: boolean;
  afterEventData: Event;
  checkClick: boolean;
  checkSpaceKeyPressed: boolean;
  checkEnterKeyPressed: boolean;
  checkOnClickWasExecuted: boolean;
  overrideId: boolean;
}
