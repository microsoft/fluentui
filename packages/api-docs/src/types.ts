/**
 * Options that define the information necessary to find the api.json file
 * and create individual page json files out of it
 */
export interface IPageJsonOptions {
  /** Input files generated by api-extractor (one per package) */
  apiJsonPaths: string[];
  /** Output \*.page.json folder path. Page groups (if used) will go in subfolders. */
  outputRoot: string;
  /** If true, the generated files won't be pretty-formatted */
  min?: boolean;
  /**
   * List of allowed page names (`@docCategory` tag values) grouped by any desired criteria,
   * such as package name.
   */
  pageGroups?: PageGroups;
  /**
   * Group for any page names (`@docCategory` tag values) not listed in `pageGroups`.
   * If unspecified, unlisted pages will go under the root.
   */
  fallbackGroup?: string;
}

/** Map from group name to page names */
export type PageGroups = { [groupName: string]: string[] };

/**
 * Structure of the page.json files
 */
export interface IPageJson {
  tables: ITableJson[];
  name: string;
  group?: string;
}

export type ApiKind = 'interface' | 'enum' | 'class' | 'typeAlias';

/**
 * Info for a table representing a top-level API item: interface, enum, class, or type alias.
 */
export interface ITableJson {
  kind: ApiKind;
  name: string;
  description?: string;
  /**
   * Any types the item extends, translated to an array of text elements and links to other types.
   * For classes and interfaces only.
   *
   * Example: `Readonly<IFoo>` might translate to:
   * `[{ text: 'Readonly<' }, { text: 'IFoo', linkedPage: 'Foo', linkedPageGroup: 'components' }, { text: '>' }]`
   */
  extendsTokens?: ILinkToken[];
  members?: ITableRowJson[] | IEnumTableRowJson[];
  deprecated?: boolean;
  deprecatedMessage?: string;
}

/**
 * Generic row for API reference tables.
 * It can represent a member (property or method) of an interface or class.
 */
export interface ITableRowJson {
  name: string;
  kind?: 'method' | 'property';
  /**
   * The row's type translated to an array of text elements and links to other types.
   * For example, `Readonly<IFoo>` might translate to:
   * `[{ text: 'Readonly<' }, { text: 'IFoo', linkedPage: 'Foo', linkedPageGroup: 'components' }, { text: '>' }]`
   */
  typeTokens: ILinkToken[];
  defaultValue?: string;
  description?: string;
  deprecated?: boolean;
  deprecatedMessage?: string;
  required?: boolean;
}

/**
 * Enum member row for API reference tables.
 */
export type IEnumTableRowJson = Omit<ITableRowJson, 'kind' | 'typeTokens' | 'defaultValue' | 'required'> & {
  value: string;
};

/**
 * Text excerpt token that is part of a type definition or extends block and may have a link
 * to another doc page.
 */
export interface ILinkToken {
  text: string;
  /** If this token is a link, name of the doc page it points to */
  linkedPage?: string;
  /** If this token is a link, group/category of the doc page it points to */
  linkedPageGroup?: string;
}
