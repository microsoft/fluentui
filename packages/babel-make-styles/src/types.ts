import type { TransformOptions } from '@babel/core';
import type { EvalRule } from '@linaria/babel-preset';

export type BabelPluginOptions = {
  /** Defines set of modules and imports handled by a plugin. */
  modules?: { moduleSource: string; importName: string }[];

  /**
   * If you need to specify custom Babel configuration, you can pass them here. These options will be used by the
   * plugin when parsing and evaluating modules.
   */
  babelOptions?: Pick<TransformOptions, 'plugins' | 'presets'>;

  /** The set of rules that defines how the matched files will be transformed during the evaluation. */
  evaluationRules?: EvalRule[];
};
