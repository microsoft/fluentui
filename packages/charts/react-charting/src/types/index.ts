export * from './IDataPoint';
export * from './ILegendDataItem';
