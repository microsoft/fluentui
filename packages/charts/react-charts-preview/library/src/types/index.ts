export * from './DataPoint';
export * from './LegendDataItem';
