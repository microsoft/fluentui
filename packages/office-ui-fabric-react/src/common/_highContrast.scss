@mixin high-contrast {
  @media screen and (-ms-high-contrast: active), screen and (forced-colors: active) {
    @content;
  }
}

@mixin high-contrast-black-on-white {
  @media screen and (-ms-high-contrast: black-on-white), screen and (forced-colors: active) and (prefers-color-scheme: light) {
   @content;
  }
}

@mixin high-contrast-white-on-black {
  @media screen and (-ms-high-contrast: white-on-black), screen and (forced-colors: active) and (prefers-color-scheme: dark) {
   @content;
  }
}

@mixin high-contrast-no-adjust {
  @media screen and (-ms-high-contrast: active), screen and (forced-colors: active) {
    -ms-high-contrast-adjust: none;
    forced-color-adjust: none;
  }
}

/**
 * The style which turns off high contrast adjustment in (only) Edge Chromium browser.
 */
@mixin high-contrast-edge-chromium-no-adjust {
  @media screen and (-ms-high-contrast: active), screen and (forced-colors: active) {
    -ms-high-contrast-adjust: none;
    forced-color-adjust: none;
  }
}
