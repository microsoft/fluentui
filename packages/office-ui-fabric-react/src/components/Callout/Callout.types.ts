import * as React from 'react';
import { IStyle, ITheme } from '../../Styling';
import { DirectionalHint } from '../../common/DirectionalHint';
import { IRectangle, IStyleFunctionOrObject } from '../../Utilities';
import { ICalloutPositionedInfo } from '../../utilities/positioning';
import { ILayerProps } from '../../Layer';
import { IPopupProps } from '../../Popup';
import { Target } from '@uifabric/react-hooks';

export { Target };

/**
 * {@docCategory Callout}
 */
export interface ICalloutProps extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * The target that the Callout should try to position itself based on.
   * It can be either an Element a querySelector string of a valid Element
   * or a MouseEvent. If MouseEvent is given then the origin point of the event will be used.
   */
  target?: Target;

  /**
   * How the element should be positioned
   * @defaultvalue DirectionalHint.BottomAutoEdge
   */
  directionalHint?: DirectionalHint;

  /**
   * How the element should be positioned in RTL layouts.
   * If not specified, a mirror of the `directionalHint` alignment edge will be used instead.
   * This means that `DirectionalHint.BottomLeft` will change to `DirectionalHint.BottomRight` but
   * `DirectionalHint.LeftAuto` will not change.
   */
  directionalHintForRTL?: DirectionalHint;

  /**
   * The gap between the Callout and the target
   * @defaultvalue 0
   */
  gapSpace?: number;

  /**
   * The width of the beak.
   * @defaultvalue 16
   */
  beakWidth?: number;

  /**
   * Custom width for callout including borders. If value is 0, no width is applied.
   * @defaultvalue 0
   */
  calloutWidth?: number;

  /**
   * Maximum width for callout including borders. If value is 0, no width is applied.
   * @defaultvalue 0
   */
  calloutMaxWidth?: number;

  /**
   * Minimum width for callout including borders. If value is 0, no width is applied.
   * @defaultvalue 0
   */
  calloutMinWidth?: number;

  /**
   * The background color of the Callout in hex format ie. #ffffff.
   * @defaultvalue $ms-color-white
   */
  backgroundColor?: string;

  /**
   * The bounding rectangle (or callback that returns a rectangle) for which  the contextual menu can appear in.
   */
  bounds?: IRectangle | ((target?: Target, targetWindow?: Window) => IRectangle | undefined);

  /**
   * The minimum distance the callout will be away from the edge of the screen.
   *  @defaultvalue 8
   */
  minPagePadding?: number;

  /**
   * If true then the beak is visible. If false it will not be shown.
   * @defaultvalue true
   */
  isBeakVisible?: boolean;

  /**
   * If true then the callout will not dismiss on scroll
   * @defaultvalue false
   * @deprecated use preventDismissOnEvent callback instead
   */
  preventDismissOnScroll?: boolean;

  /**
   * If true then the callout will not dismiss on resize
   * @defaultvalue false
   * @deprecated use preventDismissOnEvent callback instead
   */
  preventDismissOnResize?: boolean;

  /**
   * If true then the callout will not dismiss when it loses focus
   * @defaultvalue false
   * @deprecated use preventDismissOnEvent callback instead
   */
  preventDismissOnLostFocus?: boolean;

  /**
   * If true then the callout will dismiss when the target element is clicked
   * @defaultvalue false
   */
  dismissOnTargetClick?: boolean;

  /**
   * If true then the callout will dismiss when the window gets focus
   * @defaultvalue false
   */
  shouldDismissOnWindowFocus?: boolean;

  /**
   * If defined, then takes priority over preventDismissOnLostFocus, preventDismissOnResize,
   * and preventDismissOnScroll.
   * If it returns true, then callout will not dismiss for this event.
   * If not defined or returns false, callout can dismiss for this event.
   */
  preventDismissOnEvent?: (ev: Event | React.FocusEvent | React.KeyboardEvent | React.MouseEvent) => boolean;

  /**
   * If true the position returned will have the menu element cover the target.
   * If false then it will position next to the target;
   * @defaultvalue false
   */
  coverTarget?: boolean;

  /**
   * If true the positioning logic will prefer to flip edges rather than to nudge the rectangle to fit within bounds,
   * thus making sure the element aligns perfectly with target's alignment edge
   */
  alignTargetEdge?: boolean;

  /**
   * Aria role assigned to the callout (Eg. dialog, alertdialog).
   */
  role?: string;

  /**
   * Accessible label text for callout.
   */
  ariaLabel?: string;

  /**
   *  Defines the element id referencing the element containing label text for callout.
   */
  ariaLabelledBy?: string;

  /**
   * Defines the element id referencing the element containing the description for the callout.
   */
  ariaDescribedBy?: string;

  /**
   * CSS class to apply to the callout.
   * @defaultvalue null
   */
  className?: string;

  /**
   * CSS style to apply to the callout.
   *
   * If you set `overflowY` in this object, it provides a performance optimization by preventing
   * Popup (underlying component of Callout) from calculating whether it needs a scroll bar.
   */
  style?: React.CSSProperties;

  /**
   * Optional callback when the layer content has mounted.
   */
  onLayerMounted?: () => void;

  /**
   * Optional props to pass to the Layer component hosting the panel.
   */
  layerProps?: ILayerProps;

  /**
   * Optional props to pass the Popup component that the callout uses.
   */
  popupProps?: IPopupProps;

  /**
   * Optional callback that is called once the callout has been correctly positioned.
   * @param positions - Gives the user information about how the callout is positioned such as the
   * final edge of the target that it positioned against, the beak position, and the beaks relationship to the
   * edges of the callout.
   */
  onPositioned?: (positions?: ICalloutPositionedInfo) => void;

  /**
   * Callback when the Callout tries to close.
   */
  onDismiss?: (ev?: any) => void;

  /**
   * If true do not render on a new layer. If false render on a new layer.
   */
  doNotLayer?: boolean;

  /**
   * If true the position will not change sides in an attempt to fit the callout within bounds.
   * It will still attempt to align it to whatever bounds are given.
   * @defaultvalue false
   */
  directionalHintFixed?: boolean;

  /**
   * Specify the final height of the content.
   * To be used when expanding the content dynamically so that callout can adjust its position.
   */
  finalHeight?: number;

  /**
   * Manually set OverflowYHidden style prop to true on calloutMain element
   * A variety of callout load animations will need this to hide the scollbar that can appear
   */
  hideOverflow?: boolean;

  /**
   * If true then the callout will attempt to focus the first focusable element that it contains.
   * If it doesn't find an element, no focus will be set and the method will return false.
   * This means that it's the contents responsibility to either set focus or have
   * focusable items.
   * @returns True if focus was set, false if it was not.
   */
  setInitialFocus?: boolean;

  /**
   * Set max height of callout
   * When not set the callout will expand with contents up to the bottom of the screen
   */
  calloutMaxHeight?: number;

  /**
   * Callback when the Callout body is scrolled.
   */
  onScroll?: () => void;

  /**
   * Optional theme for component
   */
  theme?: ITheme;

  /**
   * Optional styles for the component.
   */
  styles?: IStyleFunctionOrObject<ICalloutContentStyleProps, ICalloutContentStyles>;

  /**
   * If specified, renders the Callout in a hidden state.
   * Use this flag, rather than rendering a callout conditionally based on visibility,
   * to improve rendering performance when it becomes visible.
   * Note: When callout is hidden its content will not be rendered. It will only render
   * once the callout is visible.
   */
  hidden?: boolean;

  /**
   * If true, the component will be updated even when hidden=true.
   * Note that this would consume resources to update even though
   * nothing is being shown to the user.
   * This might be helpful though if your updates are small and you want the
   * callout to be revealed fast to the user when hidden is set to false.
   */
  shouldUpdateWhenHidden?: boolean;

  /**
   * If true, when this component is unmounted, focus will be restored to the element that had focus when the component
   * first mounted.
   * @defaultvalue true
   * @deprecated use onRestoreFocus callback instead
   */
  shouldRestoreFocus?: boolean;

  /**
   * Called when the component is unmounting, and focus needs to be restored.
   * Argument passed down contains two variables, the element that the underlying
   * popup believes focus should go to * and whether or not the popup currently
   * contains focus. If this is provided, focus will not be restored automatically,
   * you'll need to call originalElement.focus()
   */
  onRestoreFocus?: (options: {
    originalElement?: HTMLElement | Window;
    containsFocus: boolean;
    documentContainsFocus: boolean;
  }) => void;
}

/**
 * {@docCategory Callout}
 */
export interface ICalloutContentStyleProps {
  /**
   * Theme to apply to the calloutContent.
   */
  theme: ITheme;

  /**
   * Width for callout including borders.
   */
  calloutWidth?: number;

  /**
   * CSS class to apply to the callout.
   */
  className?: string;

  /**
   * Callout positioning data
   */
  positions?: ICalloutPositionedInfo;

  /**
   * Whether or not to clip content of the callout,
   * if it overflows vertically.
   */
  overflowYHidden?: boolean;

  /**
   * Background color for the beak and callout.
   */
  backgroundColor?: string;

  /**
   * Width of Callout beak
   */
  beakWidth?: number;

  /**
   * Max width for callout including borders.
   */
  calloutMaxWidth?: number;

  /**
   * Min width for callout including borders.
   */
  calloutMinWidth?: number;
}

/**
 * {@docCategory Callout}
 */
export interface ICalloutContentStyles {
  /**
   * Style for wrapper of Callout component.
   */
  container: IStyle;

  /**
   * Style for callout container root element.
   */
  root: IStyle;

  /**
   * Style for callout beak.
   */
  beak: IStyle;

  /**
   * Style for callout beak curtain.
   */
  beakCurtain: IStyle;

  /**
   * Style for content component of the callout.
   */
  calloutMain: IStyle;
}
