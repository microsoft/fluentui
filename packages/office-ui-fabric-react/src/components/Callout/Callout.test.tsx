import * as React from 'react';
import * as ReactDOM from 'react-dom';

import * as ReactTestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';
import { Callout } from './Callout';
import { ICalloutProps } from './Callout.types';
import { CalloutContent } from './CalloutContent';
import { DirectionalHint } from '../../common/DirectionalHint';

class CalloutContentWrapper extends React.Component<ICalloutProps, {}> {
  public render(): JSX.Element {
    return <CalloutContent {...this.props} />;
  }
}

describe('Callout', () => {
  let realDom: HTMLDivElement;
  beforeEach(() => {
    realDom = document.createElement('div');
    document.body.appendChild(realDom);
  });
  afterEach(() => {
    ReactDOM.unmountComponentAtNode(realDom);
    document.body.removeChild(realDom);
    jest.useRealTimers();
    jest.resetAllMocks();
  });

  it('renders Callout correctly', () => {
    const component = renderer.create(<CalloutContentWrapper>Content</CalloutContentWrapper>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('target id strings does not throw exception', () => {
    let threwException = false;
    try {
      ReactTestUtils.renderIntoDocument<HTMLDivElement>(
        <div>
          <button id="target" style={{ top: '10px', left: '10px', height: '0', width: '0px' }}>
            {' '}
            target{' '}
          </button>
          <Callout target="#target" directionalHint={DirectionalHint.topLeftEdge}>
            <div>Content</div>
          </Callout>
        </div>,
      );
    } catch (e) {
      threwException = true;
    }

    expect(threwException).toEqual(false);
  });

  it('target MouseEvents does not throw exception', () => {
    const mouseEvent = document.createEvent('MouseEvent');
    const eventTarget = document.createElement('div');
    mouseEvent.initMouseEvent('click', false, false, window, 0, 0, 0, 0, 0, false, false, false, false, 1, eventTarget);
    let threwException = false;
    try {
      ReactTestUtils.renderIntoDocument<HTMLDivElement>(
        <div>
          <Callout target={eventTarget} directionalHint={DirectionalHint.topLeftEdge}>
            <div>Content</div>
          </Callout>
        </div>,
      );
    } catch (e) {
      threwException = true;
    }

    expect(threwException).toEqual(false);
  });

  it('target Elements does not throw exception', () => {
    const targetElement = document.createElement('div');
    document.body.appendChild(targetElement);
    let threwException = false;
    try {
      ReactTestUtils.renderIntoDocument<HTMLDivElement>(
        <div>
          <Callout target={targetElement} directionalHint={DirectionalHint.topLeftEdge}>
            <div>Content</div>
          </Callout>
        </div>,
      );
    } catch (e) {
      threwException = true;
    }

    expect(threwException).toEqual(false);
  });

  it('without target does not throw exception', () => {
    let threwException = false;
    try {
      ReactTestUtils.renderIntoDocument<HTMLDivElement>(
        <div>
          <Callout directionalHint={DirectionalHint.topLeftEdge}>
            <div>Content</div>
          </Callout>
        </div>,
      );
    } catch (e) {
      threwException = true;
    }
    expect(threwException).toEqual(false);
  });

  it('passes event to onDismiss prop', () => {
    jest.useFakeTimers();
    let threwException = false;
    let gotEvent = false;
    const onDismiss = (ev?: any) => {
      if (ev) {
        gotEvent = true;
      }
    };

    // In order to have eventlisteners that have been added to the window to be called the JSX needs
    // to be rendered into the real dom rather than the testutil simulated dom.

    try {
      ReactDOM.render<HTMLDivElement>(
        <div>
          <button id="focustarget"> button </button>
          <button id="target" style={{ top: '10px', left: '10px', height: '0', width: '0px' }}>
            {' '}
            target{' '}
          </button>
          <Callout target="#target" directionalHint={DirectionalHint.topLeftEdge} onDismiss={onDismiss}>
            <div>Content</div>
          </Callout>
        </div>,
        realDom,
      );
    } catch (e) {
      threwException = true;
    }
    expect(threwException).toEqual(false);

    const focusTarget = document.querySelector('#focustarget') as HTMLButtonElement;

    // Move focus
    jest.runAllTimers();
    focusTarget.focus();
    expect(gotEvent).toEqual(true);
  });

  it('does not dismiss when window loses focus', () => {
    jest.useFakeTimers();
    let threwException = false;
    let gotEvent = false;
    // Callout will only call the dismiss event if hasFocus returns false
    // so this needs to be mocked
    const b = jest.spyOn(window.document, 'hasFocus');
    b.mockReturnValue(false);
    const onDismiss = (ev?: any) => {
      if (ev) {
        gotEvent = true;
      }
    };

    // In order to have eventlisteners that have been added to the window to be called the JSX needs
    // to be rendered into the real dom rather than the testutil simulated dom.

    try {
      ReactDOM.render<HTMLDivElement>(
        <div>
          <button id="target" style={{ top: '10px', left: '10px', height: '0', width: '0px' }}>
            {' '}
            target{' '}
          </button>
          <Callout target="#target" directionalHint={DirectionalHint.topLeftEdge} onDismiss={onDismiss}>
            <div>Content</div>
            <button title="foo" id="blurtarget">
              text
            </button>
          </Callout>
        </div>,
        realDom,
      );
    } catch (e) {
      threwException = true;
    }
    expect(threwException).toEqual(false);

    const blurtarget = document.getElementById('blurtarget') as HTMLElement;

    // Move focus
    jest.runAllTimers();
    // Since this is a native event handler, rather than a react one, the event
    // must be triggered like this.
    blurtarget.dispatchEvent(new FocusEvent('blur', { relatedTarget: null }));
    expect(gotEvent).toEqual(false);
  });

  it('does dismiss focus when window loses focus when shouldDismissOnWindowFocus is true', () => {
    jest.useFakeTimers();
    let threwException = false;
    let gotEvent = false;
    // Callout will only call the dismiss event if hasFocus returns false
    // so this needs to be mocked.
    const b = jest.spyOn(window.document, 'hasFocus');
    b.mockReturnValue(false);
    const onDismiss = (ev?: any) => {
      if (ev) {
        gotEvent = true;
      }
    };

    // In order to have eventlisteners that have been added to the window to be called the JSX needs
    // to be rendered into the real dom rather than the testutil simulated dom.

    try {
      ReactDOM.render<HTMLDivElement>(
        <div>
          <button id="target" style={{ top: '10px', left: '10px', height: '0', width: '0px' }}>
            {' '}
            target{' '}
          </button>
          <Callout
            target="#target"
            shouldDismissOnWindowFocus={true}
            directionalHint={DirectionalHint.topLeftEdge}
            onDismiss={onDismiss}
          >
            <div>Content</div>
            <button title="foo" id="blurtarget">
              text
            </button>
          </Callout>
        </div>,
        realDom,
      );
    } catch (e) {
      threwException = true;
    }
    expect(threwException).toEqual(false);

    const blurtarget = document.getElementById('blurtarget') as HTMLElement;

    // Move focus
    jest.runAllTimers();
    // Since this is a native event handler, rather than a react one, the event
    // must be triggered like this.
    blurtarget.dispatchEvent(new FocusEvent('blur', { relatedTarget: null }));
    expect(gotEvent).toEqual(true);
  });

  it('It will correctly return focus to element that spawned it', () => {
    jest.useFakeTimers();

    const focusedElement = document.createElement('button');
    focusedElement.innerHTML = 'Inner HTML so we can be sure it is right';
    // Callout/popup checks active element to get what currently has focus
    // to know what to return focus to. By mocking the return value we can be sure
    // that it will have something "focused" when mounted
    const b = jest.spyOn(window.document, 'activeElement', 'get');
    b.mockReturnValue(focusedElement as Element);

    let threwException = false;
    let previousFocusElement;
    let isFocused;
    let restoreCalled = false;
    const onRestoreFocus = (options: { originalElement: any; containsFocus: any; documentContainsFocus: any }) => {
      previousFocusElement = options.originalElement;
      isFocused = options.containsFocus;
      restoreCalled = true;
    };
    // In order to have eventlisteners that have been added to the window to be called the JSX needs
    // to be rendered into the real dom rather than the testutil simulated dom.
    try {
      ReactDOM.render<HTMLDivElement>(
        <div>
          <button id="target" style={{ top: '10px', left: '10px', height: '0', width: '0px' }}>
            target
          </button>
          <Callout target="#target" directionalHint={DirectionalHint.topLeftEdge} onRestoreFocus={onRestoreFocus}>
            {/* must be a button to be focusable for the test*/}
            <button id={'inner'}>Content</button>
          </Callout>
        </div>,
        realDom,
      );
    } catch (e) {
      threwException = true;
    }
    expect(threwException).toEqual(false);
    const focusTarget = document.querySelector('#inner') as HTMLDivElement;

    jest.runAllTimers();

    // Make sure that focus is in the callout
    focusTarget.focus();

    // Unmounting everything is the same as dismissing the Callout. As
    // the tree is unmounted, popup will get unmounted first and the
    // onRestoreFocus method will get called
    ReactDOM.unmountComponentAtNode(realDom);

    expect(restoreCalled).toEqual(true);
    expect(isFocused).toEqual(true);

    // Just to make sure that both elements are not undefined
    expect(previousFocusElement).not.toBeFalsy();
    expect(previousFocusElement).toEqual(focusedElement);
  });

  it('passes popupProps through to Popup', () => {
    let threwException = false;

    try {
      ReactDOM.render<HTMLDivElement>(
        <div>
          <Callout doNotLayer role="dialog" popupProps={{ 'aria-modal': 'true' }}>
            content
          </Callout>
        </div>,
        realDom,
      );
    } catch (e) {
      threwException = true;
    }

    expect(threwException).toEqual(false);
    const popup = document.querySelector('[role=dialog]') as HTMLDivElement;

    expect(popup.getAttribute('aria-modal')).toEqual('true');
  });
});
