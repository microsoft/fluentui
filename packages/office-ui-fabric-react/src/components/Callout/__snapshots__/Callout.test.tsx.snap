// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Callout renders Callout correctly 1`] = `
<div
  className=
      ms-Callout-container
      {
        position: relative;
      }
  style={undefined}
>
  <div
    className=
        ms-Callout
        {
          border-color: #eaeaea;
          border-style: solid;
          border-width: 1px;
          box-shadow: 0 0 5px 0px rgba(0,0,0,0.4);
          box-sizing: border-box;
          outline: transparent;
          position: absolute;
        }
        @media screen and (-ms-high-contrast: active){& {
          border-color: WindowText;
          border-style: solid;
          border-width: 1px;
        }
        &::-moz-focus-inner {
          border: 0px;
        }
    style={
      Object {
        "filter": "opacity(0)",
        "opacity": 0,
      }
    }
    tabIndex={-1}
  >
    <div
      aria-describedby={undefined}
      aria-label={undefined}
      aria-labelledby={undefined}
      className=
          ms-Callout-main
          {
            background-color: #ffffff;
            overflow-x: hidden;
            overflow-y: auto;
            position: relative;
          }
      onKeyDown={[Function]}
      onScroll={undefined}
      role={undefined}
      style={
        Object {
          "maxHeight": 750,
          "overflowY": undefined,
        }
      }
    >
      Content
    </div>
  </div>
</div>
`;
